File: Phase03/src/main/java/InvertedIndex.java
package main.java;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;



public class InvertedIndex implements IInvertedIndex {

    final private HashMap<String, HashSet<String>> dataHashMap;

    public InvertedIndex(IFileReader fileReader){
        dataHashMap = tokenize(fileReader.getFilesContents(), fileReader.getStopWords());
    }


    private HashMap<String,HashSet<String>> tokenize(List<FileTuple> docs,List<String> stopWords){

        HashMap<String,HashSet<String>> outputMap = new HashMap<>();

        for(FileTuple doc : docs) {
            ArrayList<String> allWords = Stream.of(getNormalizedString(doc.getData()))
                                                .collect(Collectors.toCollection(ArrayList<String>::new));
            allWords.removeAll(stopWords);

            addWordsToMap(outputMap, doc, allWords);
        }
        return outputMap;
    }

    private void addWordsToMap(HashMap<String,HashSet<String>> outputMap, FileTuple doc, ArrayList<String> allWords) {
        for (String word : allWords){
            if (!outputMap.containsKey(word)) {
                outputMap.put(word, new HashSet<>());
            }
            outputMap.get(word).add(doc.getName());
        }
    }

    private String[] getNormalizedString(String doc) {
        return Arrays.stream(doc.split(" "))
                .filter(e -> e.trim().length() > 0)
                .toArray(String[]::new);
    }

    @Override
    public HashSet<String> getResultListByWord(String word){
        return dataHashMap.getOrDefault(word,new HashSet<>());
    }
}

File: Phase08/SearchEngineEfCore/SearchEngineEfCore/InvertedIndex.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace SearchEngineEfCore
{
    public class InvertedIndex : IInvertedIndex
    {
        private static IDatabase _database;
        public InvertedIndex(IFileReader fileReader,IDatabase databaseClass)
        {
            _database = databaseClass;
            Tokenize(fileReader.GetFilesContents(), fileReader.GetStopWords());
        }

        private static void Tokenize(List<FileTuple> docs, HashSet<string> stopWords)
        {
            foreach (var doc in docs)
            {
                AddWordsToDatabase(doc.DocName,doc.DocContent,stopWords);
            }
        }

        private static List<string> RemoveStopWords(string input, HashSet<string> stopWords)
        {
            var words = GetNormalizedString(input);
            return words
                .Where(word => !stopWords.Contains(word))
                .ToList();
        }

        private static void AddWordsToDatabase(string docName, string words, HashSet<string> stopWords)
        {
            var wordsList = RemoveStopWords(words,stopWords);
            _database.InsertWordsOfDoc(docName,wordsList);
        }


        private static List<string> GetNormalizedString(string input)
        {
            return input.Split(' ')
                .Distinct()
                .Where(x => !string.IsNullOrEmpty(x))
                .ToList();
        }

        public HashSet<string> GetResultListByWord(string word)
        {
            return _database.QueryListOfPostings(word)
                                        .Select(w => w.DocName)
                                        .ToHashSet();
        }
    }
}